File: research/verdict_research/features/rating_deconvolution.py
import math
from dataclasses import dataclass


@dataclass
class RatingDeconvolutionResult:
    injected_share: float
    residual_error: float


def rating_deconvolution(
    observed: list[float],
    organic_prior: list[float],
    injection_kernel: list[float],
) -> RatingDeconvolutionResult:
    if len(observed) != 5 or len(organic_prior) != 5 or len(injection_kernel) != 5:
        raise ValueError("rating deconvolution expects five bin histograms")

    diff = [k - p for k, p in zip(injection_kernel, organic_prior, strict=True)]
    residual_from_prior = [o - p for o, p in zip(observed, organic_prior, strict=True)]

    numerator = sum(d * r for d, r in zip(diff, residual_from_prior, strict=True))
    denominator = sum(d * d for d in diff)

    raw_a = 0.0 if denominator == 0 else numerator / denominator
    injected_share = min(1.0, max(0.0, raw_a))

    modeled = [p + injected_share * d for p, d in zip(organic_prior, diff, strict=True)]
    squared_errors = [(o - m) ** 2 for o, m in zip(observed, modeled, strict=True)]
    residual_error = math.sqrt(sum(squared_errors) / len(squared_errors))

    return RatingDeconvolutionResult(injected_share=injected_share, residual_error=residual_error)

File: extension/src/score/ratingDeconvolution.ts
export interface RatingDeconvolutionResult {
  injectedShare: number;
  residualError: number;
}

// SPEC.md 5.1: observed = (1 - a) * organicPrior + a * injectionKernel, over the
// five star bins. organicPrior and injectionKernel are supplied by the caller,
// this function only fits the scalar mixture weight `a` by constrained least
// squares (a clamped to [0, 1]) and reports the residual as an RMSE.
export function ratingDeconvolution(
  observed: readonly number[],
  organicPrior: readonly number[],
  injectionKernel: readonly number[],
): RatingDeconvolutionResult {
  if (observed.length !== 5 || organicPrior.length !== 5 || injectionKernel.length !== 5) {
    throw new Error("rating deconvolution expects five bin histograms");
  }

  const diff = injectionKernel.map((value, i) => value - (organicPrior[i] ?? 0));
  const residualFromPrior = observed.map((value, i) => value - (organicPrior[i] ?? 0));

  const numerator = sum(diff.map((d, i) => d * (residualFromPrior[i] ?? 0)));
  const denominator = sum(diff.map((d) => d * d));

  const rawA = denominator === 0 ? 0 : numerator / denominator;
  const injectedShare = Math.min(1, Math.max(0, rawA));

  const modeled = organicPrior.map((prior, i) => prior + injectedShare * (diff[i] ?? 0));
  const squaredErrors = observed.map((value, i) => (value - (modeled[i] ?? 0)) ** 2);
  const residualError = Math.sqrt(sum(squaredErrors) / squaredErrors.length);

  return { injectedShare, residualError };
}

function sum(values: readonly number[]): number {
  return values.reduce((total, value) => total + value, 0);
}
